### PR TITLE
feat(grouping): Do not flip stuff randomly to in_app false

### PR DESCRIPTION
**Pull request by mitsuhiko**
This also adds other changes that deal with some of the consequenes of this.  For
instance the code for chained exceptions will now remove the value of an exception
if at least one other exception has a stack.

### CODE REVIEW

**Review by jan-auer — APPROVED** [review]
Seems like the new code in `chained_exception` might have to change again once we add a component for `type`.

Apart from the snapshot changes in similarity, 👍

File: tests/sentry/grouping/similarity/snapshots/test_features/test_similarity_extract_fingerprinting/fingerprint_native.pysnap [hunk 1]
```diff
@@ -59,7 +57,6 @@ similarity:2020-07-23:value:character-5-shingle: "e hea"
 similarity:2020-07-23:value:character-5-shingle: "e sym"
 similarity:2020-07-23:value:character-5-shingle: "eader"
 similarity:2020-07-23:value:character-5-shingle: "ed to"
-similarity:2020-07-23:value:character-5-shingle: "f ran"
```
Comment (untitaker) [inline tests/sentry/grouping/similarity/snapshots/test_features/test_similarity_extract_fingerprinting/fingerprint_native.pysnap:62]: this seems suspicious

File: tests/sentry/grouping/snapshots/test_variants/test_event_hash_variant/legacy@2019_03_12/exception_compute_hashes.pysnap [hunk 2]
```diff
@@ -9,7 +9,7 @@ app:
     app (exception of system takes precedence)
       exception (ignored because hash matches system variant)
         stacktrace*
-          frame* (frame considered in-app because no frame is in-app)
+          frame*
```
Comment (untitaker) [inline tests/sentry/grouping/snapshots/test_variants/test_event_hash_variant/legacy@2019_03_12/exception_compute_hashes.pysnap:12]: doesn't matter, variant is discarded

File: tests/sentry/grouping/snapshots/test_variants/test_event_hash_variant/mobile@2021_02_12/exception_compute_hashes.pysnap [hunk 3]
```diff
@@ -1,20 +1,20 @@
...
 ---
 app-depth-max:
-  hash: null
+  hash: "9509e122c6175606d52862fa4f64853c"
   component:
-    app-depth-max (exception of system takes precedence)
-      exception (ignored because this variant does not have a contributing stacktrace, but the system variant does)
-        stacktrace
-          frame (non app frame)
+    app-depth-max*
```
Comment (untitaker) [inline tests/sentry/grouping/snapshots/test_variants/test_event_hash_variant/mobile@2021_02_12/exception_compute_hashes.pysnap:9]: this was bs behavior anyway

File: tests/sentry/grouping/snapshots/test_variants/test_event_hash_variant/newstyle@2019_04_17/java_chained.pysnap [hunk 4]
```diff
@@ -173,7 +173,7 @@ app:
                 "bind0"
           type*
             "BindException"
-          value*
+          value (ignored because one exception has a stacktrace)
```
Comment (untitaker) [inline tests/sentry/grouping/snapshots/test_variants/test_event_hash_variant/newstyle@2019_04_17/java_chained.pysnap:176]: this is a significant change? but I think we discussed this

File: tests/sentry/grouping/snapshots/test_variants/test_event_hash_variant/newstyle@2019_04_17/threads_compute_hashes.pysnap [hunk 5]
```diff
@@ -1,15 +1,15 @@
...
 source: tests/sentry/grouping/test_variants.py
 ---
 app:
   hash: null
   component:
     app (threads of system take precedence)
-      threads (ignored because this variant does not have a contributing stacktrace, but the system variant does)
-        stacktrace
-          frame (non app frame)
+      threads (ignored because hash matches system variant)
```
Comment (untitaker) [inline tests/sentry/grouping/snapshots/test_variants/test_event_hash_variant/newstyle@2019_04_17/threads_compute_hashes.pysnap:10]: this comes from my fix, however without armin's commits my fixes don't change snapshots at all

File: src/sentry/grouping/strategies/newstyle.py [hunk 6]
```diff
@@ -558,9 +558,32 @@ def chained_exception(chained_exception, context, **meta):
...
+                id="stacktrace", recursive=True, only_contributing=True
+            )
+            if stacktrace_component is not None:
+                have_stacktrace_exception = True
+            else:
+                value_component = component.get_subcomponent(id="value", only_contributing=True)
+                if value_component is not None:
+                    non_stacktrace_values.append(value_component)
+
+    # if one of the exception has a stack trace we want to ignore the values of
```
Comment (untitaker) [inline src/sentry/grouping/strategies/newstyle.py:579]: I think we should try reverting this to see if there are remaining hash changes

**Review by jan-auer — APPROVED** [review]
Diffs all look good, particularly system hashes never change. Good to merge.

**Pull request merged**